### ISSUE TITLE
Specify err/out files explicitly, for each task

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#586]
Given that we are moving towards multiple-tasks-per-slurm-job, we cannot any more rely on the SLURM err/out files for logging, since they would get cluttered by the logs of all tasks belonging to the same SLURM job.

To do: make sure that next to each task command (inside the SLURM submission script) there is a redirection of logs towards specific files.

**Comment by tcompa** [comment]
Re-open.

If we run the `test_full_workflow.py::test_failing_workflow_TaskExecutionError`, these are the artifacts folder for the slurm and grouped_slurm (AKA the new slurm) backends:

Current backend
```
-rw-rw-r-- 1 tommaso tommaso  229 Mar 23 14:53 0.args.json
-rw-rw-r-- 1 tommaso tommaso  335 Mar 23 14:53 0.err
-rw-rw-r-- 1 tommaso tommaso   46 Mar 23 14:53 0.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 0.out
-rw-rw-r-- 1 tommaso tommaso  299 Mar 23 14:53 0.slurm.2.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 0.slurm.2.out
-rw-rw-r-- 1 tommaso tommaso  883 Mar 23 14:53 0.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  482 Mar 23 14:53 1_par_0.args.json
-rw-rw-r-- 1 tommaso tommaso  779 Mar 23 14:53 1_par_0.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_0.out
-rw-rw-r-- 1 tommaso tommaso  299 Mar 23 14:53 1_par_0.slurm.3.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_0.slurm.3.out
-rw-rw-r-- 1 tommaso tommaso  916 Mar 23 14:53 1_par_0.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  482 Mar 23 14:53 1_par_1.args.json
-rw-rw-r-- 1 tommaso tommaso  779 Mar 23 14:53 1_par_1.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_1.out
-rw-rw-r-- 1 tommaso tommaso  299 Mar 23 14:53 1_par_1.slurm.4.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_1.slurm.4.out
-rw-rw-r-- 1 tommaso tommaso  916 Mar 23 14:53 1_par_1.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  482 Mar 23 14:53 1_par_2.args.json
-rw-rw-r-- 1 tommaso tommaso  779 Mar 23 14:53 1_par_2.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_2.out
-rw-rw-r-- 1 tommaso tommaso  299 Mar 23 14:53 1_par_2.slurm.5.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_par_2.slurm.5.out
-rw-rw-r-- 1 tommaso tommaso  916 Mar 23 14:53 1_par_2.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  100 Mar 23 14:53 metadata.json
-rw-rw-r-- 1 tommaso tommaso 5.7K Mar 23 14:53 workflow.log
```

New backend:
```
-rw-rw-r-- 1 tommaso tommaso  229 Mar 23 14:53 0.args.json
-rw-rw-r-- 1 tommaso tommaso  335 Mar 23 14:53 0.err
-rw-rw-r-- 1 tommaso tommaso 1.6K Mar 23 14:53 0.in.mw16ct3Tg6TA5mhzL6Ln9gAOFpyfuN4u.pickle
-rw-rw-r-- 1 tommaso tommaso   46 Mar 23 14:53 0.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 0.out
-rw-rw-r-- 1 tommaso tommaso  368 Mar 23 14:53 0.out.mw16ct3Tg6TA5mhzL6Ln9gAOFpyfuN4u.pickle
-rw-rw-r-- 1 tommaso tommaso  299 Mar 23 14:53 0.slurm.6.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 0.slurm.6.out
-rw-rw-r-- 1 tommaso tommaso  967 Mar 23 14:53 0.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso 1.8K Mar 23 14:53 1_batch_0.in.3Qie6nErWX454JHq4R1OvM69kam0qbgW.pickle
-rw-rw-r-- 1 tommaso tommaso 1.8K Mar 23 14:53 1_batch_0.in.iWfZR1Gb4zLj4tkbqzMPCPCZrcjhpu12.pickle
-rw-rw-r-- 1 tommaso tommaso 1.8K Mar 23 14:53 1_batch_0.in.rxEiwWrbfqA84av3e9zRFSSx4Iv4TBSq.pickle
-rw-rw-r-- 1 tommaso tommaso 3.5K Mar 23 14:53 1_batch_0.out.3Qie6nErWX454JHq4R1OvM69kam0qbgW.pickle
-rw-rw-r-- 1 tommaso tommaso 3.5K Mar 23 14:53 1_batch_0.out.iWfZR1Gb4zLj4tkbqzMPCPCZrcjhpu12.pickle
-rw-rw-r-- 1 tommaso tommaso 3.5K Mar 23 14:53 1_batch_0.out.rxEiwWrbfqA84av3e9zRFSSx4Iv4TBSq.pickle
-rw-rw-r-- 1 tommaso tommaso  897 Mar 23 14:53 1_batch_0.slurm.7.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 23 14:53 1_batch_0.slurm.7.out
-rw-rw-r-- 1 tommaso tommaso 2.1K Mar 23 14:53 1_batch_0.slurm.submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  100 Mar 23 14:53 metadata.json
-rw-rw-r-- 1 tommaso tommaso 4.5K Mar 23 14:53 workflow.log
```

Something is wrong, that is, we don't have single-task stderr/stdout files any more (e.g. `1_par_0.err`, or an equivalent one with a new name).

**Comment by tcompa** [comment]
Got it: the files do exist correctly on the user-side working folder, but they are not copied over because there is not any more a match in the file prefix (i.e.. `1_batch_0*` does not match `1_par_2`).
That is, we are now using two different prefixes (one for the single-task files, another one for the multi-task job), and we need to modify the matching function to include both - when copying files from user's folder to server's folder.

**Comment by tcompa** [comment]
Almost fixed with https://github.com/fractal-analytics-platform/fractal-server/commit/e700d681d715c683c73ec97f5ef8a981a1139070. EDIT: remaining small bug fixed with as of 4624df74.

The current server-side file list (for a workflow with two non-parallel tasks and then one task parallelized over three components) looks like
```
-rw-rw-r-- 1 tommaso tommaso  296 Mar 24 11:09 0.args.json
-rw-rw-r-- 1 tommaso tommaso  388 Mar 24 11:09 0.err
-rw-rw-r-- 1 tommaso tommaso 1.6K Mar 24 11:08 0_in_PkAku2UsDkyuf2DWQDKf3hBSPaOlHlMm.pickle
-rw-rw-r-- 1 tommaso tommaso   46 Mar 24 11:09 0.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 0.out
-rw-rw-r-- 1 tommaso tommaso  368 Mar 24 11:09 0_out_PkAku2UsDkyuf2DWQDKf3hBSPaOlHlMm.pickle
-rw-rw-r-- 1 tommaso tommaso  299 Mar 24 11:09 0_slurm_2.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 0_slurm_2.out
-rw-rw-r-- 1 tommaso tommaso  882 Mar 24 11:08 0_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  473 Mar 24 11:09 1.args.json
-rw-rw-r-- 1 tommaso tommaso  401 Mar 24 11:09 1.err
-rw-rw-r-- 1 tommaso tommaso 1.6K Mar 24 11:09 1_in_zL2OEEUQ7yVFhAtYN0eKWL5O1kzj3OkB.pickle
-rw-rw-r-- 1 tommaso tommaso   46 Mar 24 11:09 1.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 1.out
-rw-rw-r-- 1 tommaso tommaso  377 Mar 24 11:09 1_out_zL2OEEUQ7yVFhAtYN0eKWL5O1kzj3OkB.pickle
-rw-rw-r-- 1 tommaso tommaso  299 Mar 24 11:09 1_slurm_3.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 1_slurm_3.out
-rw-rw-r-- 1 tommaso tommaso  882 Mar 24 11:09 1_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  897 Mar 24 11:09 2_batch_0_slurm_4.err
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 2_batch_0_slurm_4.out
-rw-rw-r-- 1 tommaso tommaso 1.9K Mar 24 11:09 2_batch_0_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso 1.7K Mar 24 11:09 2_in_hFd9MTqqBAxaWSZ4ZpEWulHRrVL6vU0J.pickle
-rw-rw-r-- 1 tommaso tommaso 1.7K Mar 24 11:09 2_in_KMNxAqmiEW7aZrL0b8zIt9rXzZjCsT0H.pickle
-rw-rw-r-- 1 tommaso tommaso 1.7K Mar 24 11:09 2_in_srwHFZe1gBQ5tk86rm44HTx57TLKQ4QR.pickle
-rw-rw-r-- 1 tommaso tommaso   16 Mar 24 11:09 2_out_hFd9MTqqBAxaWSZ4ZpEWulHRrVL6vU0J.pickle
-rw-rw-r-- 1 tommaso tommaso   16 Mar 24 11:09 2_out_KMNxAqmiEW7aZrL0b8zIt9rXzZjCsT0H.pickle
-rw-rw-r-- 1 tommaso tommaso   16 Mar 24 11:09 2_out_srwHFZe1gBQ5tk86rm44HTx57TLKQ4QR.pickle
-rw-rw-r-- 1 tommaso tommaso  479 Mar 24 11:09 2_par_0.args.json
-rw-rw-r-- 1 tommaso tommaso  369 Mar 24 11:09 2_par_0.err
-rw-rw-r-- 1 tommaso tommaso    2 Mar 24 11:09 2_par_0.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 2_par_0.out
-rw-rw-r-- 1 tommaso tommaso  479 Mar 24 11:09 2_par_1.args.json
-rw-rw-r-- 1 tommaso tommaso  369 Mar 24 11:09 2_par_1.err
-rw-rw-r-- 1 tommaso tommaso    2 Mar 24 11:09 2_par_1.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 2_par_1.out
-rw-rw-r-- 1 tommaso tommaso  479 Mar 24 11:09 2_par_2.args.json
-rw-rw-r-- 1 tommaso tommaso  369 Mar 24 11:09 2_par_2.err
-rw-rw-r-- 1 tommaso tommaso    2 Mar 24 11:09 2_par_2.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Mar 24 11:09 2_par_2.out
-rw-rw-r-- 1 tommaso tommaso  152 Mar 24 11:09 metadata.json
-rw-rw-r-- 1 tommaso tommaso 1.5K Mar 24 11:09 workflow.log
```

We may improve it a bit further (by making pickle file names more meaningful), but at least all files seem to be there.

It was a bit more annoying than expected, because the combination of the user/sever separate folders and of multi-tasks job breaks some assumption on the filenames.

**Comment by tcompa** [comment]
Current version, after a few updates (e.g. b01371d3594196070207cb48e7680e34ec53d048, 591c3a1de184fdf8038f27196ad50551d53794ff), when running
```
pytest test_full_workflow.py::test_failing_workflow_TaskExecutionError[parallel-grouped_slurm]
```
(and by setting `keep_pickle_files=True` in the new `FractalSlurmExecutor`) leads to
```
-rw-rw-r-- 1 tommaso tommaso  228 Apr  5 12:00 0.args.json
-rw-rw-r-- 1 tommaso tommaso  334 Apr  5 12:00 0.err
-rw-rw-r-- 1 tommaso tommaso 1.6K Apr  5 12:00 0_in_JPDy5dRFyBBrHBuS7GXRbLp32YqRpWJE.pickle
-rw-rw-r-- 1 tommaso tommaso   46 Apr  5 12:00 0.metadiff.json
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 0.out
-rw-rw-r-- 1 tommaso tommaso  367 Apr  5 12:00 0_out_JPDy5dRFyBBrHBuS7GXRbLp32YqRpWJE.pickle
-rw-rw-r-- 1 tommaso tommaso  299 Apr  5 12:00 0_slurm_2.err
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 0_slurm_2.out
-rw-rw-r-- 1 tommaso tommaso  983 Apr  5 12:00 0_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  299 Apr  5 12:00 1_batch_000000_slurm_3.err
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_batch_000000_slurm_3.out
-rw-rw-r-- 1 tommaso tommaso 1.1K Apr  5 12:00 1_batch_000000_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  299 Apr  5 12:00 1_batch_000001_slurm_4.err
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_batch_000001_slurm_4.out
-rw-rw-r-- 1 tommaso tommaso 1.1K Apr  5 12:00 1_batch_000001_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  299 Apr  5 12:00 1_batch_000002_slurm_5.err
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_batch_000002_slurm_5.out
-rw-rw-r-- 1 tommaso tommaso 1.1K Apr  5 12:00 1_batch_000002_slurm_submit.sbatch
-rw-rw-r-- 1 tommaso tommaso  480 Apr  5 12:00 1_par_0.args.json
-rw-rw-r-- 1 tommaso tommaso  775 Apr  5 12:00 1_par_0.err
-rw-rw-r-- 1 tommaso tommaso 1.8K Apr  5 12:00 1_par_0_in_jJ1ZccS8Uhg4v9Tw9D8Z2zsWKDNvld0J.pickle
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_par_0.out
-rw-rw-r-- 1 tommaso tommaso 3.5K Apr  5 12:00 1_par_0_out_jJ1ZccS8Uhg4v9Tw9D8Z2zsWKDNvld0J.pickle
-rw-rw-r-- 1 tommaso tommaso  480 Apr  5 12:00 1_par_1.args.json
-rw-rw-r-- 1 tommaso tommaso  775 Apr  5 12:00 1_par_1.err
-rw-rw-r-- 1 tommaso tommaso 1.8K Apr  5 12:00 1_par_1_in_1ZrqVb6ENBHa12scaee5OHqemVaUf3DW.pickle
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_par_1.out
-rw-rw-r-- 1 tommaso tommaso 3.5K Apr  5 12:00 1_par_1_out_1ZrqVb6ENBHa12scaee5OHqemVaUf3DW.pickle
-rw-rw-r-- 1 tommaso tommaso  480 Apr  5 12:00 1_par_2.args.json
-rw-rw-r-- 1 tommaso tommaso  775 Apr  5 12:00 1_par_2.err
-rw-rw-r-- 1 tommaso tommaso 1.8K Apr  5 12:00 1_par_2_in_ZK8nO1iQ6Z8TUTzA1jlh1O4kJ2LNzbuU.pickle
-rw-rw-r-- 1 tommaso tommaso    0 Apr  5 12:00 1_par_2.out
-rw-rw-r-- 1 tommaso tommaso 3.5K Apr  5 12:00 1_par_2_out_ZK8nO1iQ6Z8TUTzA1jlh1O4kJ2LNzbuU.pickle
-rw-rw-r-- 1 tommaso tommaso  100 Apr  5 12:00 metadata.json
-rw-rw-r-- 1 tommaso tommaso 1.2K Apr  5 12:00 workflow.log
```

This should be closed with #582.
